File: backend/app/api/work_orders.py
"""Work Order intake endpoints (Phase 4 — Work Orders view, demand save).

HTTP surface for the Work Orders management view (GUI_DESIGN
§11.1–§11.3): listing/searching the WO list, loading Work Order
Details with demand lines, creating a Work Order with its demand
draft, saving edits, and removing demand lines. Business demand only —
production release lives on its own surface
(``app.api.production_release``) and no endpoint here can touch
QuantityFlows, PartMovements, or the current-position projection.

Routes stay thin orchestration: request schemas validate shape only
(``extra="forbid"`` — a client that submits a server-owned field such
as ``status``, ``allocated_quantity``, or ``priority_rank`` is
rejected instead of silently ignored), the Application layer owns
every business rule and the one-save-one-transaction protocol, and
the central handlers in ``app.api.errors`` translate typed failures.

Deliberate surface decisions:

- ``work_order_number`` is nullable in requests and responses: blank
  confirmed input persists NULL (the UI renders ``—``); an entered
  number travels and is stored verbatim. ``GET /work-orders?number=``
  is the exact-resolution lookup the New Work Order dialog uses to
  open an existing number instead of duplicating it; a POST that
  still carries an existing number is a 409 that creates nothing.
- One Save = one request = one transaction: ``POST /work-orders``
  saves the New Work Order dialog (header + all lines), ``PATCH
  /work-orders/{id}`` saves the Work Order Details draft (header
  edits + ``line_edits`` + ``new_lines``) — each all-or-nothing with
  its audit rows.
- ``released_quantity`` / ``remaining_quantity`` are derived per demand
  from the immutable ``RECEIVED`` Movement context: a demand may be
  released in several parts, so the release action stays available
  while ``remaining_quantity > 0`` and the server refuses anything
  beyond it. A line with any released quantity accepts a RESTRICTED
  edit: ``requested_quantity`` (never below the released or allocated
  quantity), ``due_date`` and ``job_numbers`` stay editable, while
  ``request_type`` — like the PN, which no edit carries — plus
  ``requester``, ``reason`` and ``notes`` are locked and removal stays
  refused. A ``line_edits`` entry that breaks the restriction is a 409
  that changes nothing.
- ``DELETE /work-orders/{id}/demands/{id}`` enforces the canonical
  removal rules (PROJECT_PROFILE §13, §8.2) in the backend, never
  only in the UI: a saved demand deletes only while no production
  quantity has ever been released for it, and the last demand line
  of a Work Order is never removable (a Work Order contains one or
  more demand records; nothing auto-deletes the Work Order) — either
  violation is a 409 that removes nothing; so is removing a line that
  stocked quantity has ever been allocated to (Phase 10).
- ``GET /work-orders/completed`` is the read-only completed history
  (Phase 10, GUI_DESIGN §11.5): server-side search, the Done range and
  the due outcome in the site calendar (``SITE_TIMEZONE`` — the one
  rule behind every row's ``done_date`` / ``due_outcome`` too), the
  chosen sort and keyset paging bound to it.
- ``priority_rank`` and ``allocated_quantity`` appear only in
  responses: Hot ranking (Phase 12) and allocation (Phase 10) own
  those values.
- The audit ``actor_reference`` is never client-writable: no request
  carries an actor, so audit rows from this surface stay NULL until an
  authenticated (or server-configured) identity exists (Phase 14).
"""

import datetime
from typing import Literal

from fastapi import APIRouter
from pydantic import BaseModel, ConfigDict, Field, StrictInt

from app.api.dependencies import SessionDep
from app.application import work_orders
from app.application.common import UNSET
from app.application.work_orders import WorkOrderDetail, WorkOrderSummary
from app.domain.enums import RequestType
from app.infrastructure.models import WorkOrderDemand

router = APIRouter(prefix="/api")


class WorkOrderDemandResponse(BaseModel):
    id: int
    work_order_id: int
    # Canonical uppercase PN, kept by the demand itself (no FK to the
    # optional master).
    part_number: str
    request_type: RequestType
    requested_quantity: int
    allocated_quantity: int
    due_date: datetime.date | None
    priority_rank: int | None
    job_numbers: list[str]
    requester: str | None
    reason: str | None
    notes: str | None
    # Server-derived release evidence (immutable RECEIVED Movement
    # context): true once any production quantity was released for this
    # demand. The UI renders such a line Released and restricts it to
    # the Qty/Due date/Job Numbers edit — the flag survives any reload
    # and is never a client-session guess.
    has_released_quantity: bool
    # Quantity already released for this demand, and what is left of
    # the requested quantity. A demand may be released in several
    # parts, so the UI keeps offering the release action while
    # remaining_quantity > 0; the server enforces the same cap.
    released_quantity: int
    remaining_quantity: int
    created_at: datetime.datetime
    updated_at: datetime.datetime


class WorkOrderSummaryResponse(BaseModel):
    id: int
    # NULL = internal Work Order without an external number; the UI
    # renders `—` (display-only, never persisted).
    work_order_number: str | None
    received_date: datetime.date
    due_date: datetime.date | None
    # Server-derived read status: OPEN while at least one current
    # demand has never released; RELEASED once every current demand
    # line carries release evidence (GUI_DESIGN §11.1); COMPLETED once
    # every line is fully allocated (Phase 10).
    status: str
    # The done date (Phase 10): set while the Work Order is completed.
    completed_at: datetime.datetime | None
    # The completion calendar (Phase 10, GUI_DESIGN §11.5): the done
    # DATE of `completed_at` in the site time zone, and the due outcome
    # judged on it — derived by the server, the one rule the history's
    # filter and every displayed row share. All null while active.
    done_date: datetime.date | None
    due_outcome: Literal["ON_TIME", "LATE", "NO_DUE_DATE"] | None
    days_late: int | None
    demand_line_count: int
    part_numbers: list[str]


class WorkOrderDetailResponse(BaseModel):
    id: int
    work_order_number: str | None
    received_date: datetime.date
    due_date: datetime.date | None
    status: str
    completed_at: datetime.datetime | None
    done_date: datetime.date | None
    due_outcome: Literal["ON_TIME", "LATE", "NO_DUE_DATE"] | None
    days_late: int | None
    created_at: datetime.datetime
    updated_at: datetime.datetime
    demands: list[WorkOrderDemandResponse]


class WorkOrderDemandCreateRequest(BaseModel):
    """One new demand line (Add Part flow, GUI_DESIGN §11.3)."""

    model_config = ConfigDict(extra="forbid")

    part_number: str
    # Strict: a quantity is an integer, never a coerced bool/float/text
    # — uncertain input is rejected, not reinterpreted.
    requested_quantity: StrictInt
    # Manual entry defaults to NEW in the Application layer.
    request_type: RequestType | None = None
    due_date: datetime.date | None = None
    job_numbers: list[str] = Field(default_factory=list)
    requester: str | None = None
    reason: str | None = None
    notes: str | None = None


class WorkOrderDemandEditRequest(BaseModel):
    """A partial edit of one saved demand line.

    The PN is deliberately absent: a different PN is a new line, never
    a rewrite of a saved one. Omitted fields keep their values; an
    explicit ``null`` due date is the valid "No due date" choice. Once
    the line has released quantity only ``requested_quantity``,
    ``due_date`` and ``job_numbers`` are still accepted, and the
    quantity may not fall below what is already released or allocated
    (409, nothing written).
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    request_type: RequestType | None = None
    requested_quantity: StrictInt | None = None
    due_date: datetime.date | None = None
    job_numbers: list[str] | None = None
    requester: str | None = None
    reason: str | None = None
    notes: str | None = None


class WorkOrderCreateRequest(BaseModel):
    """The New Work Order dialog's Save demand (GUI_DESIGN §11.3)."""

    model_config = ConfigDict(extra="forbid")

    # Blank/omitted persists NULL on an internal Work Order — no
    # temporary number is ever generated; an entered value is verbatim.
    work_order_number: str | None = None
    # Defaults to the current date during manual creation.
    received_date: datetime.date | None = None
    due_date: datetime.date | None = None
    lines: list[WorkOrderDemandCreateRequest]


class WorkOrderUpdateRequest(BaseModel):
    """The Work Order Details dialog's Save demand (GUI_DESIGN §11.2)."""

    model_config = ConfigDict(extra="forbid")

    # The audited Work Order Number edit (PROJECT_PROFILE §7): an
    # internal Work Order may receive its real external number later.
    work_order_number: str | None = None
    due_date: datetime.date | None = None
    line_edits: list[WorkOrderDemandEditRequest] = Field(default_factory=list)
    new_lines: list[WorkOrderDemandCreateRequest] = Field(default_factory=list)


def _summary_response(summary: WorkOrderSummary) -> WorkOrderSummaryResponse:
    work_order = summary.work_order
    outcome = work_orders.due_outcome_of(work_order)
    return WorkOrderSummaryResponse(
        id=work_order.id,
        work_order_number=work_order.work_order_number,
        received_date=work_order.received_date,
        due_date=work_order.due_date,
        status=summary.status,
        completed_at=work_order.completed_at,
        done_date=work_orders.done_date_of(work_order.completed_at),
        due_outcome=outcome.outcome if outcome is not None else None,
        days_late=outcome.days_late if outcome is not None else None,
        demand_line_count=summary.demand_line_count,
        part_numbers=summary.part_numbers,
    )


def _demand_response(demand: WorkOrderDemand, released: int) -> WorkOrderDemandResponse:
    return WorkOrderDemandResponse(
        id=demand.id,
        work_order_id=demand.work_order_id,
        part_number=demand.part_number,
        request_type=RequestType(demand.request_type),
        requested_quantity=demand.requested_quantity,
        allocated_quantity=demand.allocated_quantity,
        due_date=demand.due_date,
        priority_rank=demand.priority_rank,
        job_numbers=demand.job_numbers,
        requester=demand.requester,
        reason=demand.reason,
        notes=demand.notes,
        has_released_quantity=released > 0,
        released_quantity=released,
        remaining_quantity=max(demand.requested_quantity - released, 0),
        created_at=demand.created_at,
        updated_at=demand.updated_at,
    )


def _detail_response(detail: WorkOrderDetail) -> WorkOrderDetailResponse:
    work_order = detail.work_order
    outcome = work_orders.due_outcome_of(work_order)
    return WorkOrderDetailResponse(
        id=work_order.id,
        work_order_number=work_order.work_order_number,
        received_date=work_order.received_date,
        due_date=work_order.due_date,
        status=detail.status,
        completed_at=work_order.completed_at,
        done_date=work_orders.done_date_of(work_order.completed_at),
        due_outcome=outcome.outcome if outcome is not None else None,
        days_late=outcome.days_late if outcome is not None else None,
        created_at=work_order.created_at,
        updated_at=work_order.updated_at,
        demands=[
            _demand_response(demand, detail.released_quantities.get(demand.id, 0))
            for demand in detail.demands
        ],
    )


@router.get("/work-orders")
def list_work_orders(
    session: SessionDep, search: str | None = None, number: str | None = None
) -> list[WorkOrderSummaryResponse]:
    return [
        _summary_response(summary)
        for summary in work_orders.list_work_orders(session, search=search, number=number)
    ]


class CompletedWorkOrdersResponse(BaseModel):
    work_orders: list[WorkOrderSummaryResponse]
    # Matching completed Work Orders in the whole history.
    total: int
    # Completed Work Orders in the whole history regardless of the
    # filters — "none ever" versus "none in this range".
    history_total: int
    # Opaque keyset cursor of the last row — pass back as `cursor` for
    # the next page of the SAME sort; null when no further page can
    # exist.
    next_cursor: str | None


@router.get("/work-orders/completed")
def list_completed_work_orders(
    session: SessionDep,
    search: str | None = None,
    done_from: datetime.date | None = None,
    done_to: datetime.date | None = None,
    due_outcome: Literal["ALL", "ON_TIME", "LATE", "NO_DUE_DATE"] = "ALL",
    sort: Literal["DONE", "RECEIVED", "DUE", "NUMBER"] = "DONE",
    direction: Literal["ASC", "DESC"] = "DESC",
    cursor: str | None = None,
    limit: int = work_orders.COMPLETED_PAGE_LIMIT,
) -> CompletedWorkOrdersResponse:
    """The read-only completed history (GUI_DESIGN §11.5).

    `done_from` / `done_to` are inclusive done DATES in the site
    calendar (`SITE_TIMEZONE`), the same calendar the due outcome and
    every row's `done_date` use. Sorted server-side by `sort` /
    `direction` (Done descending by default, NULLs last, id as the
    tie-breaker) with keyset paging bound to that order.
    """
    page = work_orders.list_completed_work_orders(
        session,
        search=search,
        done_from=done_from,
        done_to=done_to,
        due_outcome=due_outcome,
        sort=sort,
        direction=direction,
        cursor=cursor,
        limit=limit,
    )
    return CompletedWorkOrdersResponse(
        work_orders=[_summary_response(summary) for summary in page.work_orders],
        total=page.total,
        history_total=page.history_total,
        next_cursor=page.next_cursor,
    )


@router.get("/work-orders/{work_order_id}")
def get_work_order(work_order_id: int, session: SessionDep) -> WorkOrderDetailResponse:
    return _detail_response(work_orders.get_work_order(session, work_order_id))


@router.post("/work-orders", status_code=201)
def create_work_order(body: WorkOrderCreateRequest, session: SessionDep) -> WorkOrderDetailResponse:
    detail = work_orders.create_work_order(
        session,
        work_order_number=body.work_order_number,
        received_date=body.received_date,
        due_date=body.due_date,
        lines=[line.model_dump() for line in body.lines],
    )
    return _detail_response(detail)


@router.patch("/work-orders/{work_order_id}")
def update_work_order(
    work_order_id: int, body: WorkOrderUpdateRequest, session: SessionDep
) -> WorkOrderDetailResponse:
    provided = body.model_dump(exclude_unset=True)
    detail = work_orders.update_work_order(
        session,
        work_order_id,
        # Omitted keeps the current value; an explicit null (or blank)
        # persists NULL — both are audited edits.
        work_order_number=provided.get("work_order_number", UNSET),
        due_date=provided.get("due_date", UNSET),
        line_edits=[edit.model_dump(exclude_unset=True) for edit in body.line_edits],
        new_lines=[line.model_dump() for line in body.new_lines],
    )
    return _detail_response(detail)


@router.delete("/work-orders/{work_order_id}/demands/{demand_id}", status_code=204)
def delete_work_order_demand(work_order_id: int, demand_id: int, session: SessionDep) -> None:
    """Remove one saved demand line (PROJECT_PROFILE §13, §8.2).

    Blocked with 409 once any quantity for the demand has been
    released, and blocked with 409 for the Work Order's last demand
    line; removal never cascades to the PartNumber master,
    QuantityFlows, PartMovements, release history, or other demand
    lines for the same PN.
    """
    work_orders.delete_work_order_demand(session, work_order_id, demand_id)

File: frontend/src/api/work-orders.ts
// Work Order intake API (Phase 4 — GUI_DESIGN §11.1–§11.3).
//
// Business demand only: listing/searching the WO list, loading Work
// Order Details with demand lines, creating a Work Order with its
// demand draft (one Save = one transaction), saving edits, and
// removing a saved demand line. Production release lives in
// `src/api/production-release.ts` — nothing here can create
// production quantity.
//
// Wire responses are the backend's snake_case schemas; this module
// maps them to the camelCase application types the views render, and
// back. Server-owned fields (`status`, `priority_rank`,
// `allocated_quantity`) are read-only here — the backend rejects any
// attempt to submit them.
//
// Production-safe: no mock data, no framework imports.

import { apiRequest } from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface WorkOrderSummary {
  id: number;
  /** Verbatim external number, or null on an internal Work Order —
   * rendered as `—` (display-only, never persisted). */
  workOrderNumber: string | null;
  /** ISO `YYYY-MM-DD`. */
  receivedDate: string;
  /** ISO `YYYY-MM-DD`, or null while the Work Order is unscheduled. */
  dueDate: string | null;
  /** Server-derived read status: `OPEN`, `RELEASED`, or `COMPLETED`
   * once every demand line is fully allocated (Phase 10). */
  status: string;
  /** The done date (`completed_at`, ISO timestamp) — set exactly while
   * the Work Order is completed; a later allocation reversal clears it
   * and returns the Work Order to the active list. */
  completedAt: string | null;
  demandLineCount: number;
  /** Canonical PNs of the demand lines, in demand order. */
  partNumbers: string[];
}

export interface WorkOrderDemand {
  id: number;
  workOrderId: number;
  /** The canonical uppercase PN — identity, kept verbatim. */
  partNumber: string;
  requestType: 'NEW' | 'MODIFY';
  requestedQuantity: number;
  /** Server-owned (Phase 10 allocation). */
  allocatedQuantity: number;
  dueDate: string | null;
  /** Server-owned (Phase 12 Hot ranking). */
  priorityRank: number | null;
  jobNumbers: string[];
  requester: string | null;
  reason: string | null;
  notes: string | null;
  /**
   * Server-derived release evidence (immutable RECEIVED Movement
   * context): true once any production quantity was released for this
   * demand — in any session, ever. The line renders Released and
   * accepts the restricted edit only (Qty never below the released or
   * allocated quantity, due date, Job Numbers — PROJECT_PROFILE §13);
   * this is never a client-session guess.
   */
  hasReleasedQuantity: boolean;
  /**
   * Quantity already released for this demand, and what is left of the
   * requested quantity. A demand may be released in several parts (20
   * of 50, then 12, then 18), so the release action stays offered
   * while `remainingQuantity > 0` — the server enforces the same cap.
   * Both are derived from Movement history: never stored, never a
   * session guess.
   */
  releasedQuantity: number;
  remainingQuantity: number;
}

export interface WorkOrderDetail {
  id: number;
  workOrderNumber: string | null;
  receivedDate: string;
  dueDate: string | null;
  status: string;
  /** The done date — see `WorkOrderSummary.completedAt`. */
  completedAt: string | null;
  demands: WorkOrderDemand[];
}

/** One new demand line of a save (Add Part flow, GUI_DESIGN §11.3). */
export interface NewDemandLine {
  partNumber: string;
  requestedQuantity: number;
  requestType: 'NEW' | 'MODIFY';
  dueDate: string | null;
  jobNumbers: string[];
  notes: string | null;
}

/**
 * A partial edit of one saved demand line. Only present fields travel:
 * an omitted field keeps its saved value, while an explicit null
 * `dueDate` is the valid "No due date" choice. The PN is deliberately
 * absent — a different PN is a new line, never a rewrite.
 */
export interface DemandLineEdit {
  id: number;
  requestType?: 'NEW' | 'MODIFY';
  requestedQuantity?: number;
  dueDate?: string | null;
  jobNumbers?: string[];
  notes?: string | null;
}

// ---------------------------------------------------------------------------
// Wire mapping
// ---------------------------------------------------------------------------

interface WorkOrderSummaryWire {
  id: number;
  work_order_number: string | null;
  received_date: string;
  due_date: string | null;
  status: string;
  completed_at: string | null;
  demand_line_count: number;
  part_numbers: string[];
}

interface WorkOrderDemandWire {
  id: number;
  work_order_id: number;
  part_number: string;
  request_type: 'NEW' | 'MODIFY';
  requested_quantity: number;
  allocated_quantity: number;
  due_date: string | null;
  priority_rank: number | null;
  job_numbers: string[];
  requester: string | null;
  reason: string | null;
  notes: string | null;
  has_released_quantity: boolean;
  released_quantity: number;
  remaining_quantity: number;
}

interface WorkOrderDetailWire {
  id: number;
  work_order_number: string | null;
  received_date: string;
  due_date: string | null;
  status: string;
  completed_at: string | null;
  demands: WorkOrderDemandWire[];
}

function toSummary(wire: WorkOrderSummaryWire): WorkOrderSummary {
  return {
    id: wire.id,
    workOrderNumber: wire.work_order_number,
    receivedDate: wire.received_date,
    dueDate: wire.due_date,
    status: wire.status,
    completedAt: wire.completed_at,
    demandLineCount: wire.demand_line_count,
    partNumbers: wire.part_numbers,
  };
}

function toDemand(wire: WorkOrderDemandWire): WorkOrderDemand {
  return {
    id: wire.id,
    workOrderId: wire.work_order_id,
    partNumber: wire.part_number,
    requestType: wire.request_type,
    requestedQuantity: wire.requested_quantity,
    allocatedQuantity: wire.allocated_quantity,
    dueDate: wire.due_date,
    priorityRank: wire.priority_rank,
    jobNumbers: wire.job_numbers,
    requester: wire.requester,
    reason: wire.reason,
    notes: wire.notes,
    hasReleasedQuantity: wire.has_released_quantity,
    releasedQuantity: wire.released_quantity,
    remainingQuantity: wire.remaining_quantity,
  };
}

function toDetail(wire: WorkOrderDetailWire): WorkOrderDetail {
  return {
    id: wire.id,
    workOrderNumber: wire.work_order_number,
    receivedDate: wire.received_date,
    dueDate: wire.due_date,
    status: wire.status,
    completedAt: wire.completed_at,
    demands: wire.demands.map(toDemand),
  };
}

function newLineBody(line: NewDemandLine): Record<string, unknown> {
  return {
    part_number: line.partNumber,
    requested_quantity: line.requestedQuantity,
    request_type: line.requestType,
    due_date: line.dueDate,
    job_numbers: line.jobNumbers,
    notes: line.notes,
  };
}

function lineEditBody(edit: DemandLineEdit): Record<string, unknown> {
  return {
    id: edit.id,
    ...(edit.requestType !== undefined
      ? { request_type: edit.requestType }
      : {}),
    ...(edit.requestedQuantity !== undefined
      ? { requested_quantity: edit.requestedQuantity }
      : {}),
    ...('dueDate' in edit ? { due_date: edit.dueDate } : {}),
    ...(edit.jobNumbers !== undefined ? { job_numbers: edit.jobNumbers } : {}),
    ...('notes' in edit ? { notes: edit.notes } : {}),
  };
}

// ---------------------------------------------------------------------------
// Calls
// ---------------------------------------------------------------------------

/**
 * Most Work Orders one active-list read returns — the SERVER's bound
 * (`work_orders.LIST_RESULT_LIMIT`), mirrored here only to word the
 * "refine your search" hint. Never a client-side slice.
 */
export const WORK_ORDER_LIST_LIMIT = 100;

/**
 * The active WO list, newest first — filtered AND bounded by the server.
 *
 * `search` travels as `?search=`, so the contains-match over the Work
 * Order Number (GUI_DESIGN §11.1) runs in the database and a match
 * outside the first page is still found. Nothing leaves the active
 * list before allocation-derived completion (Phase 10), so the client
 * never downloads the whole list in order to filter it locally.
 */
export async function listWorkOrders(
  search?: string,
): Promise<WorkOrderSummary[]> {
  const term = search?.trim() ?? '';
  const query = term ? `?search=${encodeURIComponent(term)}` : '';
  const wires = await apiRequest<WorkOrderSummaryWire[]>(
    `/api/work-orders${query}`,
  );
  return wires.map(toSummary);
}

/** The due outcome filter of the completed history (GUI_DESIGN §11.5). */
export type DueOutcome = 'ALL' | 'ON_TIME' | 'LATE' | 'NO_DUE_DATE';

/** Keyset position over `(completed_at, id)` — the server's cursor. */
export interface CompletedCursor {
  completedAt: string;
  id: number;
}

export interface CompletedWorkOrdersPage {
  workOrders: WorkOrderSummary[];
  /** Matching completed Work Orders in the WHOLE history. */
  total: number;
  /** Completed Work Orders in the whole history regardless of the
   * filters — tells "none ever" from "none in this range". */
  historyTotal: number;
  /** Continue with this cursor for the next page; null = no more. */
  nextCursor: CompletedCursor | null;
}

/** Rows one page of the completed history returns (the server's page
 * size, `work_orders.COMPLETED_PAGE_LIMIT`); `Show more` continues. */
export const COMPLETED_PAGE_LIMIT = 50;

interface CompletedWorkOrdersPageWire {
  work_orders: WorkOrderSummaryWire[];
  total: number;
  history_total: number;
  next_cursor_completed_at: string | null;
  next_cursor_id: number | null;
}

/**
 * The read-only completed history (GUI_DESIGN §11.5), newest done date
 * first. Search (WO Number, PN, Job Number), the done-date range, the
 * due outcome and the keyset paging all run on the SERVER — the history
 * is unbounded by design and never downloaded whole.
 */
export async function listCompletedWorkOrders(input: {
  search?: string;
  /** ISO timestamps: inclusive lower / exclusive upper `completed_at`. */
  doneFrom?: string | null;
  doneTo?: string | null;
  dueOutcome?: DueOutcome;
  cursor?: CompletedCursor | null;
}): Promise<CompletedWorkOrdersPage> {
  const params = new URLSearchParams();
  const term = input.search?.trim() ?? '';
  if (term) params.set('search', term);
  if (input.doneFrom) params.set('done_from', input.doneFrom);
  if (input.doneTo) params.set('done_to', input.doneTo);
  if (input.dueOutcome && input.dueOutcome !== 'ALL') {
    params.set('due_outcome', input.dueOutcome);
  }
  if (input.cursor) {
    params.set('cursor_completed_at', input.cursor.completedAt);
    params.set('cursor_id', String(input.cursor.id));
  }
  params.set('limit', String(COMPLETED_PAGE_LIMIT));
  const wire = await apiRequest<CompletedWorkOrdersPageWire>(
    `/api/work-orders/completed?${params.toString()}`,
  );
  return {
    workOrders: wire.work_orders.map(toSummary),
    total: wire.total,
    historyTotal: wire.history_total,
    nextCursor:
      wire.next_cursor_completed_at !== null && wire.next_cursor_id !== null
        ? {
            completedAt: wire.next_cursor_completed_at,
            id: wire.next_cursor_id,
          }
        : null,
  };
}

/**
 * Exact-resolution lookup of one entered WO Number (verbatim equality
 * — stored numbers are verbatim). The New Work Order dialog uses this
 * to open an existing number instead of duplicating it; null = free.
 */
export async function resolveWorkOrderNumber(
  workOrderNumber: string,
): Promise<WorkOrderSummary | null> {
  const wires = await apiRequest<WorkOrderSummaryWire[]>(
    `/api/work-orders?number=${encodeURIComponent(workOrderNumber)}`,
  );
  return wires.length > 0 ? toSummary(wires[0]) : null;
}

export async function getWorkOrder(id: number): Promise<WorkOrderDetail> {
  const wire = await apiRequest<WorkOrderDetailWire>(`/api/work-orders/${id}`);
  return toDetail(wire);
}

/** One Save = one transaction: header plus all lines (≥ 1 required). */
export async function createWorkOrder(input: {
  /** null persists NULL (internal Work Order); entered text verbatim. */
  workOrderNumber: string | null;
  receivedDate: string;
  dueDate: string | null;
  lines: NewDemandLine[];
}): Promise<WorkOrderDetail> {
  const wire = await apiRequest<WorkOrderDetailWire>('/api/work-orders', {
    method: 'POST',
    body: {
      work_order_number: input.workOrderNumber,
      received_date: input.receivedDate,
      due_date: input.dueDate,
      lines: input.lines.map(newLineBody),
    },
  });
  return toDetail(wire);
}

/** The Work Order Details Save demand — header edits + line edits +
 * new lines in one all-or-nothing transaction. */
export async function updateWorkOrder(
  id: number,
  input: {
    /**
     * Omit to keep the current number. A string travels VERBATIM (the
     * audited Work Order Number edit, PROJECT_PROFILE §7 — e.g. adding
     * the real external number to an internal Work Order); an explicit
     * null persists NULL. Never trimmed, reformatted, or padded here.
     */
    workOrderNumber?: string | null;
    /** Omit to keep; explicit null persists NULL (audited edit). */
    dueDate?: string | null;
    lineEdits: DemandLineEdit[];
    newLines: NewDemandLine[];
  },
): Promise<WorkOrderDetail> {
  const wire = await apiRequest<WorkOrderDetailWire>(`/api/work-orders/${id}`, {
    method: 'PATCH',
    body: {
      ...('workOrderNumber' in input
        ? { work_order_number: input.workOrderNumber }
        : {}),
      ...('dueDate' in input ? { due_date: input.dueDate } : {}),
      line_edits: input.lineEdits.map(lineEditBody),
      new_lines: input.newLines.map(newLineBody),
    },
  });
  return toDetail(wire);
}

/**
 * Remove one saved demand line. The backend enforces the canonical
 * rules (PROJECT_PROFILE §13, §8.2): released demand and the last
 * demand line of a Work Order answer 409 and remove nothing.
 */
export async function deleteWorkOrderDemand(
  workOrderId: number,
  demandId: number,
): Promise<void> {
  await apiRequest<void>(
    `/api/work-orders/${workOrderId}/demands/${demandId}`,
    {
      method: 'DELETE',
    },
  );
}
